File: java/programmers/src/택배배달과수거하기.java
import java.util.List;

public class 택배배달과수거하기 {
    class Solution {

        public long solution(int cap, int n, int[] deliveries, int[] pickups) {
            long answer = 0;
            int deliveryIndex=n-1;
            int pickupIndex = n-1;
            while(deliveryIndex>=0||pickupIndex>=0){
                int tmpDeliveryCap =0;
                int tmpPickupCap =0;
                int maxDeliveryIndex =-1;
                int maxPickUpIndex =-1;
                while(true){
                    if(deliveryIndex<0){
                        break;
                    }
                    if(deliveries[deliveryIndex]==0){
                        deliveryIndex--;
                        continue;
                    }
                    if(tmpDeliveryCap ==cap) break;
                    if(deliveries[deliveryIndex]>0){
                        maxDeliveryIndex = Math.max(maxDeliveryIndex,deliveryIndex);
                        if(tmpDeliveryCap+deliveries[deliveryIndex]>cap){ //다 못 가져간다면
                            deliveries[deliveryIndex] -=(cap-tmpDeliveryCap);
                            break;
                        }
                        else{ //초과하지 않는다면
                            tmpDeliveryCap +=deliveries[deliveryIndex];
                            deliveries[deliveryIndex] =0;
                            deliveryIndex--;
                        }
                    }
                }
                while(true){
                    if(pickupIndex<0){
                        break;
                    }
                    if(pickups[pickupIndex]==0){
                        pickupIndex--;
                        continue;
                    }
                    if(tmpPickupCap ==cap) break;
                    if(pickups[pickupIndex]>0){
                        maxPickUpIndex = Math.max(maxPickUpIndex,pickupIndex);
                        if(tmpPickupCap+ pickups[pickupIndex]>cap){ //다 못 가져간다면
                            pickups[pickupIndex] -=(cap-tmpPickupCap);
                            break;
                        }
                        else{ //초과하지 않는다면
                            tmpPickupCap +=pickups[pickupIndex];
                            pickups[pickupIndex] =0;
                            pickupIndex--;
                        }
                    }
                }
                answer +=(Math.max(maxPickUpIndex+1,maxDeliveryIndex+1)*2);
            }

            return answer;
        }
    }
}

File: java/baekjoon/src/택배배달과_수거하기.kt
import kotlin.math.max

class 택배배달과_수거하기 {
    class Solution {
        var boxToDeliverI = 0
        var boxToPickI = 0

        fun solution(
            cap: Int,
            n: Int,
            deliveries: IntArray,
            pickups: IntArray,
        ): Long {
            val a = Int.MAX_VALUE
            var answer: Long = 0L
            for (i in 0 until n) {
                if (deliveries[i] != 0) boxToDeliverI = i
                if (pickups[i] != 0) boxToPickI = i
            }
            if (boxToDeliverI == 0 && boxToPickI == 0) return answer

            while (boxToPickI >= 0 || boxToDeliverI >= 0) {
                val maxDi = boxToDeliverI
                val maxPi = boxToPickI
                var boxCapForDeliver = cap

                while (boxToDeliverI >= 0) {
                    if (boxCapForDeliver <= 0) break
                    // 가져갈 수 있는 박스가 더 많을 때
                    if (deliveries[boxToDeliverI] < boxCapForDeliver) {
                        boxCapForDeliver -= deliveries[boxToDeliverI]
                        deliveries[boxToDeliverI] = 0
                    }
                    // 가져가야하는 박스가 남았을 때
                    else if (deliveries[boxToDeliverI] >= boxCapForDeliver) {
                        deliveries[boxToDeliverI] -= boxCapForDeliver
                        boxCapForDeliver = 0
                    }

                    // 다음 가져가야하는 박스 위치 찾기
                    while (boxToDeliverI >= 0 && deliveries[boxToDeliverI] == 0) boxToDeliverI--
                }

                var boxCapForPick = cap

                while (boxToPickI >= 0) {
                    if (boxCapForPick <= 0) break
                    // 가져갈 수 있는 박스가 더 많을 때
                    if (pickups[boxToPickI] < boxCapForPick) {
                        boxCapForPick -= pickups[boxToPickI]
                        pickups[boxToPickI] = 0
                    } else if (pickups[boxToPickI] >= boxCapForPick) {
                        pickups[boxToPickI] -= boxCapForPick
                        boxCapForPick = 0
                    }

                    // 다음 가져가야하는 박스 위치 찾기
                    while (boxToPickI >= 0 && pickups[boxToPickI] == 0) boxToPickI--
                }

                // println(deliveries.toList())
                // println(pickups.toList())
                answer += (max(maxDi, maxPi) + 1).toLong()
            }

            return answer * 2
        }
    }
}
